### PR TITLE
1014 improve jurisdiction metadata upload

### LINKED ISSUE #1014: Improve Jurisdiction Metadata Upload
Jurisdiction_Metadata_Reveal222

### Improvements to this page

We should make it possible for the user to select the country (highest level jurisdiction) whose jurisdictions they want.  Once they select the country, they can press the button to download the csv.

- [x] Add a country select field
- [x] When button is pressed, load all jurisdictions from the jurisdictions hierarchy endpoint
- [x] Prepare the CSV and download it
- [x] Ensure errors are handled appropriately if the API cal fails

### PR DESCRIPTION
**Pull request by KipSigei**
closes #1014

### CODE REVIEW

**Comment by moshthepitt** [comment]
@KipSigei code looks really good & clean. 🍾

Small nitpick: please add docstrings?

File: src/components/forms/JurisdictionHierachyDownload/tests/index.test.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      initialValues: {
+        jurisdictions,
+      },
+      serviceClass: new mockedOpenSRPservice(),
+    };
+    submitJurisdictionHierachyForm(setSubmitting, setGlobalError, props as any, { jurisdictions });
+    await waitFor(() => {
+      expect(mockedOpenSRPservice).toHaveBeenCalledTimes(1);
+      expect(mockGrowl).toBeCalled();
+      expect(mockDownload).toBeCalled();
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionHierachyDownload/tests/index.test.tsx:57]: Can we assert in the tests that these have been called with the expected params?

Comment (KipSigei) [reply src/components/forms/JurisdictionHierachyDownload/tests/index.test.tsx]: @moshthepitt addressed via https://github.com/onaio/reveal-frontend/pull/1062/commits/e8d1de45d7a39ca5a0ac2b412053026bab40ea3a

File: src/components/forms/JurisdictionHierachyDownload/index.tsx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+import { OPENSRP_V2_SETTINGS, TEXT_CSV } from '../../../constants';
+import { downloadFile, successGrowl } from '../../../helpers/utils';
+import { OpenSRPService } from '../../../services/opensrp';
+import { RawOpenSRPHierarchy, TreeNode } from '../../../store/ducks/opensrp/hierarchies/types';
+import { generateJurisdictionTree } from '../../../store/ducks/opensrp/hierarchies/utils';
+import { LOCATION } from '../../TreeWalker/constants';
+
+export interface JurisdictionHierachyFile {
+  jurisdiction_id: string;
+  jurisdiction_label: string;
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionHierachyDownload/index.tsx:24]: I think that this needs to change to 'jurisdiction_name' which is what is used when [uploading the metadata](https://github.com/onaio/reveal-frontend/blob/fab106acd078fe4f991006d7006154b69647bf55/src/helpers/utils.tsx#L953).

File: src/components/forms/JurisdictionHierachyDownload/index.tsx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ */
+export interface JurisdictionHierachyDownloadFormProps {
+  disabledFields: string[];
+  serviceClass: OpenSRPService;
+  submitForm: (
+    setSubmitting: (isSubmitting: boolean) => void,
+    setGlobalError: (errorMessage: string) => void,
+    props: JurisdictionHierachyDownloadFormProps,
+    values: JurisdictionHierachyDownloadFormFields
+  ) => void;
```
Comment (p-netm) [inline src/components/forms/JurisdictionHierachyDownload/index.tsx:49]: lets declare this type somwhere else other than inline here. i mean like a variable and then pass it

File: src/components/forms/JurisdictionHierachyDownload/index.tsx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            <FormGroup>
+              <Field
+                required={true}
+                component={JurisdictionSelect}
+                cascadingSelect={false}
+                name={`jurisdictions.id`}
+                id={`jurisdictions-id`}
+                placeholder={SELECT_COUNTRY}
+                aria-label={SELECT_COUNTRY}
+                disabled={false}
```
Comment (p-netm) [inline src/components/forms/JurisdictionHierachyDownload/index.tsx:134]: seems like disabled is hardcoded to false, is this intentional.

File: src/constants.tsx [hunk 5]
```diff
@@ -142,7 +142,7 @@ export const OPENSRP_TEMPLATE_ENDPOINT = 'template';
 export const OPENSRP_PLANS_BY_USER_FILTER = 'plans/user';
 export const OPENSRP_V1_SETTINGS_ENDPOINT = 'settings/sync';
 export const OPENSRP_V2_SETTINGS = 'v2/settings/';
-export const OPENSRP_JURISDICTION_HIERARCHY_ENDPOINT = 'location/getHierarchy';
+export const OPENSRP_JURISDICTION_HIERARCHY_ENDPOINT = 'location/hierarchy';
 export const OPENSRP_PLAN_HIERARCHY_ENDPOINT = 'location/getHierarchy/plan';
```
Comment (p-netm) [inline src/constants.tsx:146]: lets also change this hierarchy endpoint  `location/getHierarchy/plan` => `location/hierarchy/plan`

File: src/components/forms/JurisdictionHierachyDownload/index.tsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+      </Formik>
+    </div>
+  );
+};
+
+/**
+ * JurisdictionHierarchyDownload - enables users to export jurisdiction hierarchies based on seelcted country
+ */
+const defaultProps: JurisdictionHierachyDownloadFormProps = {
+  disabledFields: [],
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionHierachyDownload/index.tsx:171]: Are we using this anywhere?

File: src/containers/pages/JurisdictionMetadata/index.tsx [hunk 7]
```diff
@@ -128,6 +133,14 @@ const JurisdictionMetadataImportView = () => {
     submitForm: formSubmit,
   };
 
+  /** props for the Jurisdiction Hierarchy form */
+  const jurisdictionMetadataDownloadFormProp: JurisdictionHierachyDownloadFormProps | any = {
```
Comment (moshthepitt) [inline src/containers/pages/JurisdictionMetadata/index.tsx:137]: If you remove the `any` here you'll see that this object does not conform to the `JurisdictionHierachyDownloadFormProps`  interface.

Comment (KipSigei) [reply src/containers/pages/JurisdictionMetadata/index.tsx]: addressed via https://github.com/onaio/reveal-frontend/pull/1062/commits/b61f7c5b3bdc3ee4ab8930041d3a9a864cc5fbbb

**Pull request merged**